File: python/inlays.py
class Inlay:
    def __init__(self, color, design=' '):
        self.color = color
        self.design = design

    def __str__(self):
        return str(self.color) + self.design

class Inlays:
    def __init__(self, frets):
        ''' frets is map of fret num to Inlay '''
        self.frets = frets

    def setColor(self, color):
        ''' set same color for all inlays '''
        for i in self.frets:
            self.frets[i].color = color

    def setChar(self, design):
        ''' set same design for all inlays '''
        for i in self.frets:
            self.frets[i].design = design

standardInlayFrets = [1, 3, 5, 7, 9, 12, 15, 17, 19, 21, 24]
topFrets = [12, 15, 17, 19, 21, 24]
bottomFrets = [1, 3, 5, 7, 9, 12]
doubleDotFrets = [12, 24]

def inlaysFromFrets(frets, inlay):
    return Inlays({f: inlay for f in frets})

class InlayPattern:
    ''' descrbes the inlays of an entire fretboard '''
    def __init__(self, allInlays):
        self.allInlays = allInlays 

    def DotsOnFirstString(color, design=None):
        firstStringInlays = Inlays({i: Inlay(color, ':' if i in doubleDotFrets else '.') for i in standardInlayFrets})
        if design:
            firstStringInlays = Inlays({i: Inlay(color, design) for i in standardInlayFrets})
        return InlayPattern({0: firstStringInlays})

    def DotsOnLastString(color, design=None):
        lastStringInlays = Inlays({i: Inlay(color, ':' if i in doubleDotFrets else '.') for i in standardInlayFrets})
        if design:
            lastStringInlays = Inlays({i: Inlay(color, design) for i in standardInlayFrets})
        return InlayPattern({-1: lastStringInlays})

    def SplitTopBottomDots(color, design=None):
        firstStringInlays = Inlays({i: Inlay(color, ':' if i in doubleDotFrets else '.') for i in bottomFrets})
        lastStringInlays = Inlays({i: Inlay(color, ':' if i in doubleDotFrets else '.') for i in topFrets})
        if design:
            firstStringInlays = Inlays({i: Inlay(color, design) for i in bottomFrets})
            lastStringInlays = Inlays({i: Inlay(color, design) for i in topFrets})
        return InlayPattern({0: firstStringInlays, -1: lastStringInlays})

File: js/inlays.js
class Inlay {

    constructor(color, design=' ') {
        this.color = color
        this.design = design
    }

    toString() {
        return this.color.toString() + this.design
    }
}

class Inlays {
    constructor(notes) {
        //  notes is map of fret num to Inlay // 
        this.notes = notes
    }

    setColor(color) {
        //  set same color for all inlays // 
        for (var i in this.notes) {
            this.notes[i].color = color
        }
    }

    setChar(design) {
        //  set same design for all inlays // 
        for (var i in this.notes) {
            this.notes[i].design = design
        }
    }
}

standardInlayNotes = [3, 5, 7, 9, 12, 15, 17, 19, 21, 24]
topNotes = [12, 15, 17, 19, 21, 24]
bottomNotes = [3, 5, 7, 9, 12]
doubleDotNotes = [12, 24]

function inlaysFromNotes(notes, inlay) {
    inlays = {}
    for (f in notes) {
      inlays[f] = inlay
    }
    return new Inlays(inlays)
}

class InlayPattern {
    //  descrbes the inlays of an entire fretboard // 
    constructor(allInlays) {
        this.allInlays = allInlays 
    }
}

function dotInlays(color, fretNums, design=null) {
    var inlays = {}
    fretNums.forEach(function(i) {
        if (design==null) {
            inlays[i] = new Inlay(color, doubleDotNotes.includes(i) ? ':' : '.')
        } else {
            inlays[i] = new Inlay(color, design)
        }
    })
    return new Inlays(inlays)
}

function dotsOnFirstString(color, design=null) {
    firstStringInlays = dotInlays(color, standardInlayNotes, design)
    return new InlayPattern({'0': firstStringInlays})
}

function dotsOnLastString(color, design=null) {
    lastStringInlays = dotInlays(color, standardInlayNotes, design)
    return new InlayPattern({'-1': lastStringInlays})
}

function splitTopBottomDots(color, design=null) {
    firstStringInlays = dotInlays(color, bottomNotes, design)
    lastStringInlays = dotInlays(color, topNotes, design)
    return new InlayPattern({'0': firstStringInlays, '-1': lastStringInlays})
}

